### DISCUSSION TITLE
Multi-custom domains in front-end?

### DISCUSSION THREAD

**Discussion by costas90** [ory/kratos#1907] [Q&A] [answered]
Hello,

How can we make multi-domain front-ends work with Kratos?

In our SPA app, we allow our users to add their own custom domains via CNAME records e.g. `planner.domainA.com` and  `planner.domainB.com`. This way they can use our app as a white label on their sites for their customers to use.

Auth is something we do not offer right now, but we want to add an Auth solution in which they would visit `planner.domainA.com/login` to login etc.

What is the easiest solution to achieve this?

Basically, we are looking for a solution similar to Shopify where each user can setup their own shop and auth for each shop.

Do we need a multitenancy approach? Can we automate creating new tenants?

Is there a way to make this work with one front-end domain e.g. `auth.sameDomainAsKratos.com/domainA` for logging in the user and then redirecting them to the custom domain e.g. `planner.domainA.com`? While keeping the session in the front-end.

**Comment by aeneasr** [comment, marked as answer]
Are users isolated between domain A and domain B? So can I sign up for domain A and domain B individually? Or is it all the same users? Also, does logging in into domain A mean I am logged in into domain B?

**Reply by costas90** [reply]
Users do not have to be isolated if this is the easy solution. I care a lot about choosing the easy implementation. 

User isolation would be nice but not the most important right now. Not sure how much more work this is though.

Staying logged-in in both domains is not a problem or priority. More likely, the user will probably log-in to 1-3 seperate domains in their lifetime.

What kind of implementation is required for each case?

**Reply by aeneasr** [reply]
For full isolation you will need three separate instances of Ory Kratos, or running Ory Kratos in [Ory Cloud](https://www.ory.sh) which has easy support for multi-tenancy.

If you don't need isolation, and you don't need every user to be signed in into all domains (so sign in into domain A and then immediately also be signed into domain B), then you can run the same Ory Kratos instance for these multiple domains.

**Reply by dantman** [reply]
@aeneasr I have a question about the second case in regards to WebAuthn. IIRC WebAuthn stuff is tied to the origin. So if domainA.tld and domainB.tld used the same users in the same Ory Kratos instance on different domains. Then wouldn't that mean a user's passwordless login or multi-factor key setup on domainA.tld wouldn't work to sign in to the same user on domainB.tld?

**Reply by aeneasr** [reply]
Good point - I'm not sure, can you maybe check if that is the case?

**Reply by dantman** [reply]
@dantman Sorry, evaluating Kratos for upcoming projects. I don't have a multi-domain setup yet.

But I have read the WebAuthn documentation. The key potential issue I remember is this:

https://webauthn.guide/#registration
> rp: This stands for “relying party”; it can be considered as describing the organization responsible for registering and authenticating the user. The id must be a subset of the domain currently in the browser. For example, a valid id for this page is webauthn.guide. [Read the spec.](https://w3c.github.io/webauthn/#dom-publickeycredentialcreationoptions-rp)

Which if my understanding is correct means multi-origin auth only works with WebAuthn when you do the login and user management on a shared domain. Which for Ory is only available if you use Ory Hydra and use an OAuth client instead of multiple domains using the same Kratos.
